src/data/transform.ts

```diff
@@ -0,0 +1,555 @@
+          )
+        | ((
+            | {
+                type: "text";
+                text: string;
+              }
+            | {
+                type?: "text";
+                id: ExcalidrawTextElement["id"];
+                text: string;
+              }
+          ) &
+            Partial<ExcalidrawTextElement>)
+      ) &
+        MarkOptional<ElementConstructorOpts, "x" | "y">;
+} & Partial<ExcalidrawLinearElement>;
+
+export type ValidContainer =
+  | {
+      type: Exclude<ExcalidrawGenericElement["type"], "selection">;
+      id?: ExcalidrawGenericElement["id"];
+      label?: {
+        text: string;
+        fontSize?: number;
+        fontFamily?: FontFamilyValues;
+        textAlign?: TextAlign;
+        verticalAlign?: VerticalAlign;
+      } & MarkOptional<ElementConstructorOpts, "x" | "y">;
+    } & ElementConstructorOpts;
+
+export type ExcalidrawProgrammaticElement =
+  | Extract<
+      ExcalidrawElement,
+      | ExcalidrawSelectionElement
+      | ExcalidrawFreeDrawElement
+      | ExcalidrawFrameElement
+    >
+  | ({
+      type: Extract<ExcalidrawLinearElement["type"], "line">;
+      x: number;
+      y: number;
+    } & Partial<ExcalidrawLinearElement>)
+  | ValidContainer
+  | ValidLinearElement
+  | ({
+      type: "text";
+      text: string;
+      x: number;
+      y: number;
+      id?: ExcalidrawTextElement["id"];
+    } & Partial<ExcalidrawTextElement>)
+  | ({
+      type: Extract<ExcalidrawImageElement["type"], "image">;
+      x: number;
+      y: number;
+      fileId: FileId;
+    } & Partial<ExcalidrawImageElement>);
+
+export interface ExcalidrawProgrammaticAPI {
+  elements?: readonly ExcalidrawProgrammaticElement[] | null;
+}
+export const ELEMENTS_SUPPORTING_PROGRAMMATIC_API = [
+  "rectangle",
+  "ellipse",
+  "diamond",
+  "text",
+  "arrow",
+  "line",
+  "image",
+];
+
+const DEFAULT_LINEAR_ELEMENT_PROPS = {
+  width: 300,
+  height: 0,
+};
+
+const DEFAULT_DIMENSION = 100;
+
+const bindTextToContainer = (
+  containerProps: ValidContainer | ({ type: "arrow" } & ValidLinearElement),
+  textProps: { text: string } & MarkOptional<ElementConstructorOpts, "x" | "y">,
+) => {
+  let container;
+  if (containerProps.type === "arrow") {
+    const width = containerProps.width || DEFAULT_LINEAR_ELEMENT_PROPS.width;
+    const height = containerProps.height || DEFAULT_LINEAR_ELEMENT_PROPS.height;
+    container = newLinearElement({
+      width,
+      height,
+      endArrowhead: "arrow",
+      points: [
+        [0, 0],
+        [width, height],
+      ],
+      ...containerProps,
+    });
+  } else {
+    container = newElement({
+      ...containerProps,
+    });
+  }
+  const textElement: ExcalidrawTextElement = newTextElement({
+    x: 0,
+    y: 0,
+    textAlign: TEXT_ALIGN.CENTER,
+    verticalAlign: VERTICAL_ALIGN.MIDDLE,
+    ...textProps,
+    containerId: container.id,
+    strokeColor: textProps.strokeColor || container.strokeColor,
+  });
+
+  mutateElement(container, {
```

Comment:
let's use `Object.assign` instead. `mutateElement` has side effects we don't want. We'll soon move `mutateElement` into the `Scene` class for this reason.